### ISSUE TITLE
Failure in generic code on ARM64

### ISSUE THREAD

**Issue by dsyme** [mono/mono#11263]
## Description

We’ve got a problem with Xamarin/Mono F# code on ARM64.

I’m guessing will come down to be a general problem with structs and codegen for generic code.   

I’ve verified it’s not problem with tailcalls since suppressing “tail.” generation doesn’t fix the issue.
 
The generated F# code passes peverify correctly so I’m pretty sure the bug is in Mono ARM64 support.
 
Issue: https://github.com/Microsoft/visualfsharp/issues/5765#issuecomment-430137670

 
## Steps to Reproduce

Repro ZIP: https://github.com/Microsoft/visualfsharp/files/2481919/BugTest.zip
 
1.	Set Android project as startup 
2.	Build and run Debug on ARM 64 device (e.g. Google Pixel 2) – runs OK
3.	Build and run Release on ARM 64 device (e.g. Google Pixel 2) – fails

## On which platforms did you notice this

[x] Android

**Comment by lewurm** [comment]
I thought it's weird that it only happens on the Release build, but then I saw that in your Debug configuration you have:
```xml
<AndroidSupportedAbis>armeabi,armeabi-v7a,x86</AndroidSupportedAbis>
```
while in your Release configuration you have:
```xml
<AndroidSupportedAbis>armeabi;armeabi-v7a;x86;x86_64;arm64-v8a</AndroidSupportedAbis>
```

armeabi/armeabi-v7a is 32bit, and runs on arm64 devices as well. However, if arm64 is available, Andriod will take that instead on such a device. So yes: It looks like a bug in the arm64 backend. Thank you for the repro.

**Comment by dsyme** [comment]
@lewurm Oh wow, I'd never have noticed that (the repro was provided by someone else).  So does the bug repro for you?  Many thanks!

**Comment by lewurm** [comment]
couldn't reproduce with a recent version of mono on arm64/linux.

we'll try an older mono version next; based on the XA version it was reproduce with (`9.0.0.19 (HEAD/a8a3b0ec7)`), it's this mono version: https://github.com/mono/mono/commit/4fe3280bba14cb9e81e2480bfdc6d934bcbc5c3c

as a last resort we'll have to debug it on Android 🙁

**Comment by lewurm** [comment]
can't reproduce it on arm64/linux with 4fe3280, but I can reproduce it on an Android arm64 device with Xamarin.Android 9.0.0.20

**Comment by lewurm** [comment]
```
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x9
```
remaining crash log is useless unfortunately 🙁

**Comment by dsyme** [comment]
Note there is presumably a workaround for the user which is to use

```
<AndroidSupportedAbis>armeabi,armeabi-v7a,x86</AndroidSupportedAbis>
```
in release mode, correct?
thanks

**Comment by dsyme** [comment]
BTW does it matter if generic code sharing is on/off or tweaked in some way?

**Comment by dsyme** [comment]
> It doesn't crash with a newer Xamarin.Android version anymore (Version: 9.1.199.8). In mono terms that is a jump from 2018-02 -> 2018-06, so I assume the issue you are seeing was fixed between those two releases.
> 
> Is upgrading Xamarin.Android a viable solution for you?

Thanks, this is enough the mark the issue as resolved, that's great to know!

**Comment by dsyme** [comment]
Thanks, I closed https://github.com/Microsoft/visualfsharp/issues/5765#issuecomment-432700296

**Comment by cmeeren** [comment]
Happy to hear this has been resolved! Any idea when the fix will make it into the tooling included by VS?

**Comment by lewurm** [comment]
@dsyme using `armeabi,armeabi-v7a` is an okay workaround too. It has a perf penalty on newer devices though.

@cmeeren I tried the VS beta channel, which contains Xamarin.Android 9.1.0.38, but that version still crashes 🙁 

Here are links to packages from a recent master of Xamarin.Android; don't use it in production etc. (consider the `armeabi`/`armeabi-v7a` workaround instead):

* macOS: https://bosstoragemirror.azureedge.net/wrench/monodroid-mavericks-master/d2/d20057bb73c1bfd264b1a9f4b2a504f4ef7705b8/xamarin.android-9.1.199-8.pkg
* Windows: https://bosstoragemirror.azureedge.net/wrench/monodroid-mavericks-master/d2/d20057bb73c1bfd264b1a9f4b2a504f4ef7705b8/Xamarin.Android.Sdk.9.1.199.8.vsix

**Comment by cmeeren** [comment]
Thanks. As far as I understand, the issue is fixed in 9.1.199.8, so I'll keep watch for that version in VS release notes.
